vpr/src/place/analytic_placer.cpp

```diff
@@ -410,8 +410,29 @@ void AnalyticPlacer::setup_solve_blks(t_logical_block_type_ptr blkTypes) {
  * when formulating the matrix equations), an update for members is necessary
  */
 void AnalyticPlacer::update_macros() {
+    size_t max_x = g_vpr_ctx.device().grid.width();
+    size_t max_y = g_vpr_ctx.device().grid.height();
     for (auto& macro : g_vpr_ctx.mutable_placement().pl_macros) {
         ClusterBlockId head_id = macro.members[0].blk_index;
+        bool mac_can_be_placed = true;
+        //check all members' locations are within the chip, otherwise macro can not be placed with current head position
+        for (auto member = ++macro.members.begin(); member != macro.members.end(); ++member) {
+            auto member_loc = blk_locs[head_id].loc + member->offset;
+            if (member_loc.x >= int(max_x) || member_loc.x < 0) { //member location is not on the chip
+                mac_can_be_placed = false;
+                break;
+            }
+            if (member_loc.y >= int(max_y) || member_loc.y < 0) { //member location is not on the chip
+                mac_can_be_placed = false;
+                break;
+            }
+        }
+        //if macro can not be placed in this head pos, change the head pos
```

Comment:
1. Is this always safe or could the new macro head location also cause a problem?  If there's a chance it could cause a problem, it seems best to check if the macro can be placed again after you make this change.  
2. The code would be easier to read and the routine to check if a macro can be placed would be more reusable if you extracted the "mac_can_be_placed" logic into another helper function (can be called macro_can_be_placed).  Now checking if a macro can be placed after you adjust the head macro position is trivial (one more call).
3. There may also be an existing "macro_can_be_placed" routine -- worth doing a search of the placement code to see. @MohamedElgammal might also know.

Reply:
1. I think it is safe but I can investigate more, and add this if necessary.
2. A function named "macro_can_be_placed" exists in the initial_placement.cpp. However, it also considers placement constraints which is not yet supported in Analytical Placement (I think). Should I go ahead and move the function to some file that is visible to both inital_placement.cpp and analytic_placer.cpp  or write a new one visible only to analytic_placer.cpp?
@vaughnbetz